File: flutter-tela-pokemon/lib/theme/pokemon_types.dart
import 'package:flutter/material.dart';

import 'app_colors.dart';

// Reexportado para que quem importa o tema receba também a paleta base.
export 'app_colors.dart';

/// Rótulo em português e cor de um tipo de Pokémon.
class PokemonTypeInfo {
  const PokemonTypeInfo(this.label, this.color);

  final String label;
  final Color color;
}

/// Mesma tabela usada pelo aplicativo principal, mantendo a identidade visual
/// entre as duas tecnologias.
const Map<String, PokemonTypeInfo> kPokemonTypes = <String, PokemonTypeInfo>{
  'normal': PokemonTypeInfo('Normal', Color(0xFFA8A77A)),
  'fighting': PokemonTypeInfo('Lutador', Color(0xFFC22E28)),
  'flying': PokemonTypeInfo('Voador', Color(0xFFA98FF3)),
  'poison': PokemonTypeInfo('Venenoso', Color(0xFFA33EA1)),
  'ground': PokemonTypeInfo('Terrestre', Color(0xFFE2BF65)),
  'rock': PokemonTypeInfo('Pedra', Color(0xFFB6A136)),
  'bug': PokemonTypeInfo('Inseto', Color(0xFFA6B91A)),
  'ghost': PokemonTypeInfo('Fantasma', Color(0xFF735797)),
  'steel': PokemonTypeInfo('Aço', Color(0xFFB7B7CE)),
  'fire': PokemonTypeInfo('Fogo', Color(0xFFEE8130)),
  'water': PokemonTypeInfo('Água', Color(0xFF6390F0)),
  'grass': PokemonTypeInfo('Grama', Color(0xFF63BC5A)),
  'electric': PokemonTypeInfo('Elétrico', Color(0xFFF7D02C)),
  'psychic': PokemonTypeInfo('Psíquico', Color(0xFFF95587)),
  'ice': PokemonTypeInfo('Gelo', Color(0xFF96D9D6)),
  'dragon': PokemonTypeInfo('Dragão', Color(0xFF6F35FC)),
  'dark': PokemonTypeInfo('Sombrio', Color(0xFF705746)),
  'fairy': PokemonTypeInfo('Fada', Color(0xFFD685AD)),
};

const PokemonTypeInfo _fallbackType =
    PokemonTypeInfo('Desconhecido', Color(0xFF68A090));

PokemonTypeInfo typeInfo(String typeName) =>
    kPokemonTypes[typeName] ?? _fallbackType;

/// Escolhe texto claro ou escuro sobre uma cor de fundo, para que tipos claros
/// como Elétrico e Gelo continuem legíveis.
Color contrastingTextColor(Color background) {
  return background.computeLuminance() > 0.62
      ? AppColors.text
      : Colors.white;
}

File: react-native-pokemon-explorer/src/theme/pokemonTypes.js
/**
 * Tradução e cor de cada tipo de Pokémon.
 *
 * A PokéAPI devolve os tipos sempre em inglês (types[].type.name). Este mapa é a
 * única fonte de verdade para exibi-los em português e para pintar os chips,
 * cabeçalhos e filtros com a cor correspondente.
 */
export const POKEMON_TYPES = {
  normal: { label: 'Normal', color: '#A8A77A' },
  fighting: { label: 'Lutador', color: '#C22E28' },
  flying: { label: 'Voador', color: '#A98FF3' },
  poison: { label: 'Venenoso', color: '#A33EA1' },
  ground: { label: 'Terrestre', color: '#E2BF65' },
  rock: { label: 'Pedra', color: '#B6A136' },
  bug: { label: 'Inseto', color: '#A6B91A' },
  ghost: { label: 'Fantasma', color: '#735797' },
  steel: { label: 'Aço', color: '#B7B7CE' },
  fire: { label: 'Fogo', color: '#EE8130' },
  water: { label: 'Água', color: '#6390F0' },
  grass: { label: 'Grama', color: '#63BC5A' },
  electric: { label: 'Elétrico', color: '#F7D02C' },
  psychic: { label: 'Psíquico', color: '#F95587' },
  ice: { label: 'Gelo', color: '#96D9D6' },
  dragon: { label: 'Dragão', color: '#6F35FC' },
  dark: { label: 'Sombrio', color: '#705746' },
  fairy: { label: 'Fada', color: '#D685AD' },
  stellar: { label: 'Estelar', color: '#40B5A5' },
  unknown: { label: 'Desconhecido', color: '#68A090' },
};

const FALLBACK = { label: 'Desconhecido', color: '#68A090' };

/** Ordem em que os tipos aparecem na régua de filtros da listagem. */
export const TYPE_FILTER_ORDER = [
  'grass',
  'fire',
  'water',
  'electric',
  'poison',
  'flying',
  'bug',
  'normal',
  'fighting',
  'ground',
  'rock',
  'psychic',
  'ice',
  'ghost',
  'dragon',
  'dark',
  'steel',
  'fairy',
];

export function getTypeInfo(typeName) {
  if (!typeName) return FALLBACK;
  return POKEMON_TYPES[typeName] || FALLBACK;
}

export function getTypeColor(typeName) {
  return getTypeInfo(typeName).color;
}

export function getTypeLabel(typeName) {
  return getTypeInfo(typeName).label;
}

/**
 * Decide entre texto claro ou escuro sobre uma cor de fundo.
 * Sem isso, tipos claros como Elétrico e Gelo ficariam ilegíveis com texto branco.
 */
export function getContrastingTextColor(hexColor) {
  const hex = String(hexColor).replace('#', '');
  if (hex.length !== 6) return '#FFFFFF';

  const r = parseInt(hex.slice(0, 2), 16) / 255;
  const g = parseInt(hex.slice(2, 4), 16) / 255;
  const b = parseInt(hex.slice(4, 6), 16) / 255;

  // Luminância relativa simplificada (ITU-R BT.709).
  const luminance = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luminance > 0.62 ? '#1C1C1E' : '#FFFFFF';
}
